numpy/core/src/multiarray/descriptor.c

```diff
@@ -287,7 +287,7 @@ _convert_from_tuple(PyObject *obj)
             type->elsize = itemsize;
         }
     }
-    else if (PyDict_Check(val) || PyDictProxy_Check(val)) {
+    else if (type->metadata && (PyDict_Check(val) || PyDictProxy_Check(val))) {
```

Comment:
This is fine, but I just want to note for posterity that this whole if-block is weird undocumented behavior. (as are a number of things related to metadata). 

(Actually, conceivably the "correct" behavior here if `type->metadata == NULL` might be to do `type->metadata = val`, eg compare to [the end of `convert_from_dict`](https://github.com/numpy/numpy/blob/e287741d60205bb920dea15d4e70178453db2788/numpy/core/src/multiarray/descriptor.c#L1234). But that's not clear.).

In the future we might consider removing the block anyway, so I'm fine with this as-is.